packages/aft/lib/src/constraints_checker.dart

```diff
@@ -0,0 +1,367 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'package:aft/aft.dart';
+import 'package:aws_common/aws_common.dart';
+import 'package:collection/collection.dart';
+import 'package:pub_semver/pub_semver.dart';
+import 'package:pubspec_parse/pubspec_parse.dart';
+
+typedef MismatchedDependency = ({
+  PackageInfo package,
+  String dependencyName,
+  String message,
+});
+
+sealed class ConstraintsChecker {
+  ConstraintsChecker(this.action);
+
+  /// The constraints action being performed.
+  final ConstraintsAction action;
+
+  /// Checks the constraints of [package] against the rules for this
+  /// checker.
+  ///
+  /// Returns `true` if all constraint checks pass and `false` otherwise.
+  bool checkConstraints(PackageInfo package);
+
+  /// Populated by [checkConstraints] if a constraint check fails.
+  List<MismatchedDependency> get mismatchedDependencies;
+
+  /// Processes the [expectedConstraint] for [dependencyPath] in [package].
+  ///
+  /// If [action] is `apply` or `update`, the constraint is updated when there's
+  /// a mismatch. Otherwise, if it is `check`, the mismatched constraint is recorded
+  /// as an error in [mismatchedDependencies].
+  bool _processConstraint({
+    required PackageInfo package,
+    required List<String> dependencyPath,
+    required VersionConstraint expectedConstraint,
+    required String errorMessage,
+  }) {
+    switch (action) {
+      case ConstraintsAction.check:
+        mismatchedDependencies.add(
+          (
+            package: package,
+            dependencyName: dependencyPath.last,
+            message: errorMessage,
+          ),
+        );
+      case ConstraintsAction.apply:
+      case ConstraintsAction.update:
+        package.pubspecInfo.pubspecYamlEditor.update(
```

Comment:
When a package update is automatically updated, do we print that out anywhere to inform the user like we do for mistmatchedDependencies?

Reply:
No, but since the changes would be evident, I'm not sure it adds much value.